src/canisters/frontend/ic-certified-assets/src/state_machine.rs

```diff
@@ -29,25 +35,152 @@ pub const BATCH_EXPIRY_NANOS: u64 = 300_000_000_000;
+) -> Vec<String> {
+    let mut encoding_order: Vec<String> = ENCODING_CERTIFICATION_ORDER
+        .iter()
+        .map(|enc| enc.to_string())
+        .collect();
+    encoding_order.append(
+        &mut actual_encodings
+            .filter(|encoding| !ENCODING_CERTIFICATION_ORDER.contains(&encoding.as_str()))
+            .map(|s| s.into())
+            .collect(),
+    );
+    encoding_order
+}
 
 /// The file to serve if the requested file wasn't found.
 const INDEX_FILE: &str = "/index.html";
 
 /// Default aliasing behavior.
 const DEFAULT_ALIAS_ENABLED: bool = true;
 
+const STATUS_CODES_TO_CERTIFY: [u16; 2] = [200, 304];
+
 const DEFAULT_MAX_COMPUTE_EVIDENCE_ITERATIONS: u16 = 20;
 
-type AssetHashes = RbTree<Key, Hash>;
 type Timestamp = Int;
 
 #[derive(Default, Clone, Debug, CandidType, Deserialize)]
 pub struct AssetEncoding {
     pub modified: Timestamp,
     pub content_chunks: Vec<RcBytes>,
     pub total_length: usize,
+    /// Valid as-is for v2.
+    /// For v1, also make sure that encoding name == asset.most_important_encoding_v1()
     pub certified: bool,
     pub sha256: [u8; 32],
+    pub certificate_expression: Option<CertificateExpression>,
+    pub response_hashes: Option<HashMap<u16, [u8; 32]>>,
+}
+
+impl AssetEncoding {
+    fn asset_hash_path_v2(
+        &self,
+        AssetPath(path): &AssetPath,
+        status_code: u16,
+    ) -> Option<HashTreePath> {
+        self.certificate_expression.as_ref().and_then(|ce| {
+            self.response_hashes.as_ref().and_then(|hashes| {
+                hashes.get(&status_code).and_then(|response_hash| {
+                    let mut path: Vec<NestedTreeKey> = path
+                        .into_iter()
+                        .map(|segment| segment.as_str().into())
+                        .collect();
+                    path.insert(0, "http_expr".into());
+                    path.push("<$>".into()); // asset path terminator
+                    path.push(ce.hash.as_slice().into());
+                    path.push("".into()); // no request certification - use empty node
+                    path.push(response_hash.as_slice().into());
+                    Some(path.into())
+                })
+            })
+        })
+    }
+
+    fn not_found_hash_path(&self) -> Option<HashTreePath> {
+        self.certificate_expression.as_ref().and_then(|ce| {
+            self.response_hashes
+                .as_ref()
+                .and_then(|hashes| hashes.get(&200))
+                .and_then(|response_hash| {
+                    Some(HashTreePath::from(Vec::<NestedTreeKey>::from([
+                        "http_expr".into(),
+                        "<*>".into(), // 404 not found wildcard segment
+                        ce.hash.as_slice().into(),
+                        "".into(), // no request certification - use empty node
+                        response_hash.as_slice().into(),
+                    ])))
+                })
```

Comment:
Maybe we need to investigate why clippy isn't catching this, or maybe Marcin just fixed this in [#3051](https://github.com/dfinity/sdk/pull/3051), but isn't `.and_then(|_|Some(..))` equivalent to `.map()`?
```suggestion
                .map(|response_hash| {
                    HashTreePath::from(Vec::<NestedTreeKey>::from([
                        "http_expr".into(),
                        "<*>".into(), // 404 not found wildcard segment
                        ce.hash.as_slice().into(),
                        "".into(), // no request certification - use empty node
                        response_hash.as_slice().into(),
                    ]))
                })
```